src/order/galois_connection.lean

```diff
@@ -324,3 +355,171 @@ def lift_complete_lattice [complete_lattice α] (gi : galois_insertion l u) : co
+  choice_eq := λ b h, rfl }
+
+/-- A constructor for a Galois coinsertion with the trivial `choice` function. -/
+def galois_coinsertion.monotone_intro {α β : Type*} [preorder α] [preorder β] {l : α → β} {u : β → α}
+  (hu : monotone u) (hl : monotone l) (hlu : ∀ b, l (u b) ≤ b) (hul : ∀ a, u (l a) = a) :
+  galois_coinsertion l u :=
+galois_coinsertion.of_dual (galois_insertion.monotone_intro hl.order_dual hu.order_dual hlu hul)
+
+/-- Make a `galois_coinsertion l u` from a `galois_connection l u` such that `∀ b, b ≤ l (u b)` -/
+def galois_connection.to_galois_coinsertion {α β : Type*} [preorder α] [preorder β]
+  {l : α → β} {u : β → α} (gc : galois_connection l u) (h : ∀ a, u (l a) ≤ a) :
+  galois_coinsertion l u :=
+{ choice := λ x _, u x,
+  gc := gc,
+  u_l_le := h,
+  choice_eq := λ _ _, rfl }
+
+/-- Lift the bottom along a Galois connection -/
+def galois_connection.lift_order_top {α β : Type*} [partial_order α] [order_top β]
+  {l : α → β} {u : β → α} (gc : galois_connection l u) :
+  order_top α :=
+{ top    := u ⊤,
+  le_top := assume b, gc.le_u $ le_top,
+  .. ‹partial_order α› }
+
+namespace galois_coinsertion
+
+variables {l : α → β} {u : β → α}
+
+lemma u_l_eq [partial_order α] [preorder β] (gi : galois_coinsertion l u) (a : α) :
+  u (l a) = a :=
+gi.dual.l_u_eq a
+
+lemma u_surjective [partial_order α] [preorder β] (gi : galois_coinsertion l u) :
+  surjective u :=
+gi.dual.l_surjective
+
+lemma l_injective [partial_order α] [preorder β] (gi : galois_coinsertion l u) :
+  injective l :=
+gi.dual.u_injective
+
+lemma u_inf_l [semilattice_inf α] [semilattice_inf β] (gi : galois_coinsertion l u) (a b : α) :
+  u (l a ⊓ l b) = a ⊓ b :=
+gi.dual.l_sup_u a b
+
+lemma u_infi_l [complete_lattice α] [complete_lattice β] (gi : galois_coinsertion l u)
+  {ι : Sort x} (f : ι → α) :
+  u (⨅ i, l (f i)) = ⨅ i, (f i) :=
+gi.dual.l_supr_u _
+
+lemma u_infi_of_lu [complete_lattice α] [complete_lattice β] (gi : galois_coinsertion l u)
+  {ι : Sort x} (f : ι → β) (hf : ∀ i, l (u (f i)) = f i) :
+  u (⨅ i, (f i)) = ⨅ i, u (f i) :=
+gi.dual.l_supr_of_ul _ hf
+
+lemma u_sup_l [semilattice_sup α] [semilattice_sup β] (gi : galois_coinsertion l u) (a b : α) :
+  u (l a ⊔ l b) = a ⊔ b :=
+gi.dual.l_inf_u _ _
+
+lemma u_supr_l [complete_lattice α] [complete_lattice β] (gi : galois_coinsertion l u)
+  {ι : Sort x} (f : ι → α) :
+  u (⨆ i, l (f i)) = ⨆ i, (f i) :=
+gi.dual.l_infi_u _
+
+lemma u_supr_of_lu [complete_lattice α] [complete_lattice β] (gi : galois_coinsertion l u)
+  {ι : Sort x} (f : ι → β) (hf : ∀ i, l (u (f i)) = f i) :
+  u (⨆ i, (f i)) = ⨆ i, u (f i) :=
+gi.dual.l_infi_of_ul _ hf
+
+lemma l_le_l_iff [preorder α] [preorder β] (gi : galois_coinsertion l u) {a b} :
+  l a ≤ l b ↔ a ≤ b :=
+gi.dual.u_le_u_iff
+
+lemma strict_mono_l [partial_order α] [preorder β] (gi : galois_coinsertion l u) : strict_mono l :=
+λ a b h, gi.dual.strict_mono_u h
+
+section lift
+
+variables [partial_order α]
+
+/-- Lift the infima along a Galois coinsertion -/
```

Comment:
Should this be "infimum" instead of "infima" (here and below)?

Reply:
The word infimum doesn't just refer to the function, it refers to the output of the function, so each set has it's own infimum, so I'm lifting all of the infima. I don't feel strongly either way about this, but I've left it as it is.